### PR TITLE
feat(app-cmds): support nsfw only commands

### PR DESCRIPTION
**Pull request by Soheab**
## Summary

This PR adds support for marking an Application command as NSFW or not-18 or Age Restricted.

This PR also adds the missing types in `nextcord/types/interactions.py` as seen in the [Application Commands API documention](https://discord.com/developers/docs/interactions/application-commands#application-command-object)

## Checklist

- [x] If code changes were made then they have been tested.
  - [x] I have updated the documentation to reflect the changes.
  - [x] I have run `task pyright` and fixed the relevant issues.
- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

**Comment by lol219** [comment]
Finally

**Review by lol219 — APPROVED** [review]

**Comment by lol219** [comment]
@ooliver1 can you take a look of the impl

**Comment by ooliver1** [comment]
i already get notifications from new PRs btw

**Comment by lol219** [comment]
> i already get notifications from new PRs btw

Apologies

**Comment by ooliver1** [comment]
Added the stalled label as this is waiting on a relevant discord/discord-api-docs PR

**Comment by lol219** [comment]
> Added the stalled label as this is waiting on a relevant discord/discord-api-docs PR

@ooliver1 ok

File: nextcord/application_command.py [hunk 1]
```diff
@@ -1558,6 +1559,9 @@ def __init__(
             Permission(s) required to use the command. Inputting ``8`` or ``Permissions(administrator=True)`` for
             example will only allow Administrators to use the command. If set to 0, nobody will be able to use it by
             default. Server owners CAN override the permission requirements.
+        nsfw: :class:`bool`
+            If the command should be usable in NSFW channels or not. Setting to ``True`` will disable the command from
+            being usable in Not-NSFW channels.
```
Comment (Skelmis) [inline nextcord/application_command.py:1564]:
```
Non NSFW
```

File: nextcord/application_command.py [hunk 2]
```diff
@@ -2497,6 +2514,9 @@ def __init__(
...
```
Comment (Skelmis) [inline nextcord/application_command.py:2519]: Same here

File: nextcord/application_command.py [hunk 3]
```diff
@@ -2571,6 +2593,9 @@ def __init__(
...
```
Comment (Skelmis) [inline nextcord/application_command.py:2598]: Same here

File: nextcord/application_command.py [hunk 4]
```diff
@@ -2647,6 +2674,11 @@ def slash_command(
         Permission(s) required to use the command. Inputting ``8`` or ``Permissions(administrator=True)`` for
         example will only allow Administrators to use the command. If set to 0, nobody will be able to use it by
         default. Server owners CAN override the permission requirements.
+    nsfw: :class:`bool`
+        If the command should be usable in NSFW channels or not. Setting to ``True`` will disable the command from
+        being usable in Not-NSFW channels.
```
Comment (Skelmis) [inline nextcord/application_command.py:2679]: Same here

File: nextcord/application_command.py [hunk 5]
```diff
@@ -2700,6 +2734,9 @@ def message_command(
...
```
Comment (Skelmis) [inline nextcord/application_command.py:2739]: Same here

File: nextcord/application_command.py [hunk 6]
```diff
@@ -2751,6 +2790,9 @@ def user_command(
...
```
Comment (Skelmis) [inline nextcord/application_command.py:2795]: Same here

File: nextcord/client.py [hunk 7]
```diff
@@ -2447,6 +2448,9 @@ def user_command(
             Permission(s) required to use the command. Inputting ``8`` or ``Permissions(administrator=True)`` for
             example will only allow Administrators to use the command. If set to 0, nobody will be able to use it by
             default. Server owners CAN override the permission requirements.
+        nsfw: :class:`bool`
+            If the command should be usable in NSFW channels or not. Setting to ``True`` will disable the command from
+            being usable in Not-NSFW channels.
```
Comment (Skelmis) [inline nextcord/client.py:2453]: Same here

File: nextcord/client.py [hunk 8]
```diff
@@ -2494,6 +2500,9 @@ def message_command(
...
```
Comment (Skelmis) [inline nextcord/client.py:2505]: Same here

File: nextcord/client.py [hunk 9]
```diff
@@ -2549,6 +2559,11 @@ def slash_command(
...
```
Comment (Skelmis) [inline nextcord/client.py:2564]: Same here

File: nextcord/application_command.py (hunk 1 above)
Comment (searchify7) [reply nextcord/application_command.py:1564]: Non-NSFW, more better I think

**Comment by searchify7** [comment]
Hmm

File: nextcord/application_command.py [hunk 10]
```diff
@@ -1771,6 +1776,9 @@ def get_payload(self, guild_id: Optional[int]) -> dict:
                 #  the default (True) to ensure payload parity for comparisons.
                 ret["dm_permission"] = True
 
+        if self.nsfw:
```
Comment (ooliver1) [inline nextcord/application_command.py:1779]: might as well use `bool` over `Optional[bool]` as `None` and `False` are removed here anyway

**Review by ooliver1 — CHANGES_REQUESTED** [review]

File: nextcord/application_command.py (hunk 10 above)
Comment (searchify7) [reply nextcord/application_command.py:1779]: But it's not really an issue.

Comment (ooliver1) [reply nextcord/application_command.py:1779]: i didnt say it was, it was just a suggestion

Comment (searchify7) [reply nextcord/application_command.py:1779]: I see

Comment (Soheab) [reply nextcord/application_command.py:1781]: Stalled per [this message](https://canary.discord.com/channels/881118111967883295/990633056207523860/991811066822676592) in [this thread](https://canary.discord.com/channels/881118111967883295/990633056207523860) on the [Nextcord server](https://discord.com/invite/nextcord)

File: nextcord/client.py [hunk 11]
```diff
@@ -2447,6 +2448,11 @@ def user_command(
...
+
+            .. versionadded:: 2.0
```
Comment (ooliver1) [inline nextcord/client.py:2455]:
```suggestion
            .. versionadded:: 2.1
```

File: nextcord/client.py [hunk 12]
```diff
@@ -2494,6 +2502,11 @@ def message_command(
...
```
Comment (ooliver1) [inline nextcord/client.py:2509]:
```suggestion
            .. versionadded:: 2.1
```

File: nextcord/client.py [hunk 13]
```diff
@@ -2549,6 +2563,13 @@ def slash_command(
             Permission(s) required to use the command. Inputting ``8`` or ``Permissions(administrator=True)`` for
             example will only allow Administrators to use the command. If set to 0, nobody will be able to use it by
             default. Server owners CAN override the permission requirements.
+        nsfw: :class:`bool`
+            If the command should be usable in NSFW channels or not. Setting to ``True`` will disable the command from
+            being usable in Non NSFW channels.
+            .. note::
+                Due to a discord limitation, this can only be set for the parent command for subcommands.
+
+            .. versionadded:: 2.0
```
Comment (ooliver1) [inline nextcord/client.py:2572]:
```suggestion
            .. versionadded:: 2.1
```

**Review by ooliver1 — CHANGES_REQUESTED** [review]

File: nextcord/application_command.py [hunk 14]
```diff
@@ -2624,6 +2636,13 @@ def __init__(
             Permission(s) required to use the command. Inputting ``8`` or ``Permissions(administrator=True)`` for
             example will only allow Administrators to use the command. If set to 0, nobody will be able to use it by
             default. Server owners CAN override the permission requirements.
+        nsfw: :class:`bool`
+            If the command should be usable in NSFW channels or not. Setting to ``True`` will disable the command from
```
Comment (EmreTech) [inline nextcord/application_command.py:2640]: This seems confusing. The name of the parameter is `nsfw`, so I would expect it to enable support in nsfw channels. But it doesn't?

File: nextcord/types/interactions.py [hunk 15]
```diff
@@ -46,20 +46,26 @@ class _ApplicationCommandOptional(TypedDict, total=False):
     type: ApplicationCommandType
     guild_id: Snowflake
     options: List[ApplicationCommandOption]
-    default_permission: bool
+    dm_permission: bool
+    nsfw: bool
+    default_permission: bool  # This is deprecated, remove it in the future.
+    name_localizations: Dict[str, str]
+    description_localizations: Dict[str, str]
```
Comment (EmreTech) [inline nextcord/types/interactions.py:53]: Don't these localization and permission changes belong in a separate PR?

Comment (Soheab) [reply nextcord/types/interactions.py:53]: Yes but I asked @alentoghostflame in a thread called ["Support nsfw kwarg for application command"](https://discord.com/channels/881118111967883295/990633056207523860/990658931737129040) in the [Nextcord server](https://discord.com/invite/nextcord) and she said it was fine 🤷‍♀️

Comment (EmreTech) [reply nextcord/types/interactions.py:53]: I see

File: nextcord/application_command.py (hunk 14 above)
Comment (Soheab) [reply nextcord/application_command.py:2640]: Yeah, all of the descriptions need a rework...

**Comment by EmreTech** [comment]
Looks like the feature this PR is implementing will be released next Wednesday: [Message from the Discord Developers server](https://canary.discord.com/channels/613425648685547541/697138785317814292/1040031507026284645)

File: nextcord/application_command.py [hunk 16]
```diff
@@ -1812,6 +1813,11 @@ def __init__(
...
+            being usable in Non NSFW channels.
+
+            .. versionadded:: 2.1
```
Comment (EmreTech) [inline nextcord/application_command.py:1820]:
```suggestion
            .. versionadded:: 2.4
```
I don't think this PR will be able to make it in 2.3

File: nextcord/application_command.py [hunk 17]
```diff
@@ -2624,6 +2636,13 @@ def __init__(
             Permission(s) required to use the command. Inputting ``8`` or ``Permissions(administrator=True)`` for
             example will only allow Administrators to use the command. If set to 0, nobody will be able to use it by
             default. Server owners CAN override the permission requirements.
+        nsfw: :class:`bool`
+            If the command should be usable in NSFW channels or not. Setting to ``True`` will disable the command from
+            being usable in Non NSFW channels.
+            .. note::
+                Due to a discord limitation, this can only be set for the parent command for subcommands.
+
+            .. versionadded:: 2.1
```
Comment (EmreTech) [inline nextcord/application_command.py:2645]:
```suggestion
            .. versionadded:: 2.4
```

File: nextcord/application_command.py [hunk 18]
```diff
@@ -2781,6 +2802,11 @@ def __init__(
             Permission(s) required to use the command. Inputting ``8`` or ``Permissions(administrator=True)`` for
             example will only allow Administrators to use the command. If set to 0, nobody will be able to use it by
             default. Server owners CAN override the permission requirements.
+        nsfw: :class:`bool`
+            If the command should be usable in NSFW channels or not. Setting to ``True`` will disable the command from
+            being usable in Non NSFW channels.
+
+            .. versionadded:: 2.1
```
Comment (EmreTech) [inline nextcord/application_command.py:2809]:
```suggestion
            .. versionadded:: 2.4
```

File: nextcord/application_command.py [hunk 19]
```diff
@@ -2855,6 +2883,11 @@ def __init__(
...
```
Comment (EmreTech) [inline nextcord/application_command.py:2890]:
```suggestion
            .. versionadded:: 2.4
```

File: nextcord/application_command.py [hunk 20]
```diff
@@ -2931,6 +2966,13 @@ def slash_command(
         Permission(s) required to use the command. Inputting ``8`` or ``Permissions(administrator=True)`` for
         example will only allow Administrators to use the command. If set to 0, nobody will be able to use it by
         default. Server owners CAN override the permission requirements.
+    nsfw: :class:`bool`
+        If the command should be usable in NSFW channels or not. Setting to ``True`` will disable the command from
+        being usable in Non NSFW channels.
+        .. note::
```
Comment (EmreTech) [inline nextcord/application_command.py:2972]:
```suggestion
        being usable in Non NSFW channels.

        .. note::
```

File: nextcord/application_command.py [hunk 21]
```diff
@@ -2931,6 +2966,13 @@ def slash_command(
...
+            Due to a discord limitation, this can only be set for the parent command for subcommands.
+
+        .. versionadded:: 2.1
```
Comment (EmreTech) [inline nextcord/application_command.py:2975]:
```suggestion
        .. versionadded:: 2.4
```

File: nextcord/application_command.py [hunk 22]
```diff
@@ -2984,6 +3028,11 @@ def message_command(
         Permission(s) required to use the command. Inputting ``8`` or ``Permissions(administrator=True)`` for
         example will only allow Administrators to use the command. If set to 0, nobody will be able to use it by
         default. Server owners CAN override the permission requirements.
+    nsfw: :class:`bool`
+        If the command should be usable in NSFW channels or not. Setting to ``True`` will disable the command from
+        being usable in Non NSFW channels.
+
+        .. versionadded:: 2.1
```
Comment (EmreTech) [inline nextcord/application_command.py:3035]:
```suggestion
        .. versionadded:: 2.4
```

File: nextcord/application_command.py [hunk 23]
```diff
@@ -3035,6 +3086,11 @@ def user_command(
...
```
Comment (EmreTech) [inline nextcord/application_command.py:3093]:
```suggestion
        .. versionadded:: 2.4
```

File: nextcord/client.py [hunk 24]
```diff
@@ -2502,6 +2503,11 @@ def user_command(
             Permission(s) required to use the command. Inputting ``8`` or ``Permissions(administrator=True)`` for
             example will only allow Administrators to use the command. If set to 0, nobody will be able to use it by
             default. Server owners CAN override the permission requirements.
+        nsfw: :class:`bool`
+            If the command should be usable in NSFW channels or not. Setting to ``True`` will disable the command from
+            being usable in Not-NSFW channels.
+
+            .. versionadded:: 2.1
```
Comment (EmreTech) [inline nextcord/client.py:2510]:
```suggestion
            .. versionadded:: 2.4
```

File: nextcord/client.py [hunk 25]
```diff
@@ -2549,6 +2557,11 @@ def message_command(
...
```
Comment (EmreTech) [inline nextcord/client.py:2564]:
```suggestion
            .. versionadded:: 2.4
```

File: nextcord/client.py [hunk 26]
```diff
@@ -2604,6 +2618,13 @@ def slash_command(
             Permission(s) required to use the command. Inputting ``8`` or ``Permissions(administrator=True)`` for
             example will only allow Administrators to use the command. If set to 0, nobody will be able to use it by
             default. Server owners CAN override the permission requirements.
+        nsfw: :class:`bool`
+            If the command should be usable in NSFW channels or not. Setting to ``True`` will disable the command from
+            being usable in Non NSFW channels.
+            .. note::
```
Comment (EmreTech) [inline nextcord/client.py:2624]:
```suggestion
            being usable in Non NSFW channels.

            .. note::
```

File: nextcord/client.py [hunk 27]
```diff
@@ -2604,6 +2618,13 @@ def slash_command(
...
+                Due to a discord limitation, this can only be set for the parent command for subcommands.
+
+            .. versionadded:: 2.1
```
Comment (EmreTech) [inline nextcord/client.py:2627]:
```suggestion
            .. versionadded:: 2.4
```

**Review by EmreTech — CHANGES_REQUESTED** [review]

File: nextcord/application_command.py (hunk 16 above)
Comment (ooliver1) [reply nextcord/application_command.py:1915]: It isn't even guaranteed discord will release this before 2.4.

**Comment by lol219** [comment]
Yes Pls

**Comment by ooliver1** [comment]
> Yes Pls

Please use reactions instead of making unnecessary comments.

**Comment by lol219** [comment]
For the ext.application_checks, can we change the behaviour of `is_nsfw` to use that kwarg

**Comment by kingmigdor** [comment]
> > For the ext.application_checks, can we change the behaviour of `is_nsfw` to use that kwarg
> 
> if it can't keep its old implementation, i'd rather remove it.
> NSFW commands is already hidden in normal channels, what's the point of keeping it? 
> 
> we can (maybe?) keep this for legacy reason. or keep it deprecated and remove it in a (near) future version. idk

It would be a breaking change if you'd remove it. I also feel like changing its behaviour would be better rather than just removing it

**Comment by Soheab** [comment]
> > > For the ext.application_checks, can we change the behaviour of `is_nsfw` to use that kwarg
> > 
> > 
> > if it can't keep its old implementation, i'd rather remove it.
> > NSFW commands is already hidden in normal channels, what's the point of keeping it?
> > we can (maybe?) keep this for legacy reason. or keep it deprecated and remove it in a (near) future version. idk
> 
> It would be a breaking change if you'd remove it. I also feel like changing its behaviour would be better rather than just removing it

I don't think we should change the behaviour of the check or do anything with it since it would...
1. Be a breaking change
2. Some people might not want the command to be hidden and only want a library side check.
3. Did you see 1?

**Comment by ooliver1** [comment]
The whole point of `application_checks` is client-side checking. We could have removed `*has_*permissions` when `default_member_permissions` was released, but thats not what the module is for.

File: nextcord/application_command.py (hunk 10 above)
Comment (EmreTech) [reply nextcord/application_command.py:2146]: Since Discord publicly announced the release date of this feature, I think it's best to unstall this review comment.

File: nextcord/application_command.py [hunk 28]
```diff
@@ -1914,6 +1915,11 @@ def __init__(
             Permission(s) required to use the command. Inputting ``8`` or ``Permissions(administrator=True)`` for
             example will only allow Administrators to use the command. If set to 0, nobody will be able to use it by
             default. Server owners CAN override the permission requirements.
+        nsfw: :class:`bool`
+            If the command should be usable in NSFW channels or not. Setting to ``True`` will disable the command from
+            being usable in Non NSFW channels.
```
Comment (EmreTech) [inline nextcord/application_command.py:1920]:
```suggestion
            If the command should be usable in NSFW channels or not. Setting to ``True`` will make the command
            only usable in NSFW channels.
```

File: nextcord/application_command.py [hunk 29]
```diff
@@ -2730,6 +2742,13 @@ def __init__(
...
```
Comment (EmreTech) [inline nextcord/application_command.py:2747]:
```suggestion
            If the command should be usable in NSFW channels or not. Setting to ``True`` will make the command
            only usable in NSFW channels.
```

File: nextcord/application_command.py [hunk 30]
```diff
@@ -2730,6 +2742,13 @@ def __init__(
...
+            .. note::
+                Due to a discord limitation, this can only be set for the parent command for subcommands.
```
Comment (EmreTech) [inline nextcord/application_command.py:2749]:
```suggestion

            .. note::

                Due to a discord limitation, this can only be set for the parent command for subcommands.

```

File: nextcord/application_command.py [hunk 31]
```diff
@@ -2889,6 +2910,11 @@ def __init__(
             Permission(s) required to use the command. Inputting ``8`` or ``Permissions(administrator=True)`` for
             example will only allow Administrators to use the command. If set to 0, nobody will be able to use it by
             default. Server owners CAN override the permission requirements.
+        nsfw: :class:`bool`
+            If the command should be usable in NSFW channels or not. Setting to ``True`` will disable the command from
+            being usable in Non NSFW channels.
```
Comment (EmreTech) [inline nextcord/application_command.py:2915]:
```suggestion
            If the command should be usable in NSFW channels or not. Setting to ``True`` will make the command
            only usable in NSFW channels.
```

File: nextcord/application_command.py [hunk 32]
```diff
@@ -2965,6 +2993,11 @@ def __init__(
...
```
Comment (EmreTech) [inline nextcord/application_command.py:2998]:
```suggestion
            If the command should be usable in NSFW channels or not. Setting to ``True`` will make the command
            only usable in NSFW channels.
```

File: nextcord/application_command.py [hunk 33]
```diff
@@ -3043,6 +3078,13 @@ def slash_command(
         Permission(s) required to use the command. Inputting ``8`` or ``Permissions(administrator=True)`` for
         example will only allow Administrators to use the command. If set to 0, nobody will be able to use it by
         default. Server owners CAN override the permission requirements.
+    nsfw: :class:`bool`
+        If the command should be usable in NSFW channels or not. Setting to ``True`` will disable the command from
+        being usable in Non NSFW channels.
```
Comment (EmreTech) [inline nextcord/application_command.py:3083]:
```suggestion
        If the command should be usable in NSFW channels or not. Setting to ``True`` will make the command
        only usable in NSFW channels.
```

File: nextcord/application_command.py [hunk 34]
```diff
@@ -3043,6 +3078,13 @@ def slash_command(
...
+        .. note::
+            Due to a discord limitation, this can only be set for the parent command for subcommands.
```
Comment (EmreTech) [inline nextcord/application_command.py:3085]:
```suggestion
        .. note::

            Due to a discord limitation, this can only be set for the parent command for subcommands.
```

File: nextcord/application_command.py [hunk 35]
```diff
@@ -3098,6 +3142,11 @@ def message_command(
         Permission(s) required to use the command. Inputting ``8`` or ``Permissions(administrator=True)`` for
         example will only allow Administrators to use the command. If set to 0, nobody will be able to use it by
         default. Server owners CAN override the permission requirements.
+    nsfw: :class:`bool`
+        If the command should be usable in NSFW channels or not. Setting to ``True`` will disable the command from
+        being usable in Non NSFW channels.
```
Comment (EmreTech) [inline nextcord/application_command.py:3147]:
```suggestion
        If the command should be usable in NSFW channels or not. Setting to ``True`` will make the command
        only usable in NSFW channels.
```

File: nextcord/application_command.py [hunk 36]
```diff
@@ -3151,6 +3202,11 @@ def user_command(
...
```
Comment (EmreTech) [inline nextcord/application_command.py:3207]:
```suggestion
        If the command should be usable in NSFW channels or not. Setting to ``True`` will make the command
        only usable in NSFW channels.
```

File: nextcord/client.py [hunk 37]
```diff
@@ -2611,6 +2612,11 @@ def user_command(
             Permission(s) required to use the command. Inputting ``8`` or ``Permissions(administrator=True)`` for
             example will only allow Administrators to use the command. If set to 0, nobody will be able to use it by
             default. Server owners CAN override the permission requirements.
+        nsfw: :class:`bool`
+            If the command should be usable in NSFW channels or not. Setting to ``True`` will disable the command from
+            being usable in Not-NSFW channels.
```
Comment (EmreTech) [inline nextcord/client.py:2617]:
```suggestion
            If the command should be usable in NSFW channels or not. Setting to ``True`` will make the command
            only usable in NSFW channels.
```

File: nextcord/client.py [hunk 38]
```diff
@@ -2660,6 +2668,11 @@ def message_command(
...
```
Comment (EmreTech) [inline nextcord/client.py:2673]:
```suggestion
            If the command should be usable in NSFW channels or not. Setting to ``True`` will make the command
            only usable in NSFW channels.
```

File: nextcord/client.py [hunk 39]
```diff
@@ -2717,6 +2731,13 @@ def slash_command(
             Permission(s) required to use the command. Inputting ``8`` or ``Permissions(administrator=True)`` for
             example will only allow Administrators to use the command. If set to 0, nobody will be able to use it by
             default. Server owners CAN override the permission requirements.
+        nsfw: :class:`bool`
+            If the command should be usable in NSFW channels or not. Setting to ``True`` will disable the command from
+            being usable in Non NSFW channels.
```
Comment (EmreTech) [inline nextcord/client.py:2736]:
```suggestion
            If the command should be usable in NSFW channels or not. Setting to ``True`` will make the command
            only usable in NSFW channels.
```

File: nextcord/client.py [hunk 40]
```diff
@@ -2717,6 +2731,13 @@ def slash_command(
...
+            .. note::
+                Due to a discord limitation, this can only be set for the parent command for subcommands.
```
Comment (EmreTech) [inline nextcord/client.py:2738]:
```suggestion
            .. note::

                Due to a discord limitation, this can only be set for the parent command for subcommands.
```

**Review by EmreTech — CHANGES_REQUESTED** [review]
I'm going to apply all of these suggestions for you

File: nextcord/application_command.py [hunk 41]
```diff
@@ -2730,6 +2742,13 @@ def __init__(
             Permission(s) required to use the command. Inputting ``8`` or ``Permissions(administrator=True)`` for
             example will only allow Administrators to use the command. If set to 0, nobody will be able to use it by
             default. Server owners CAN override the permission requirements.
+        nsfw: :class:`bool`
+            If the command should be usable in NSFW channels or not. Setting to ``True`` will make the command
+            only usable in NSFW channels.
+            .. note::
+                Due to a discord limitation, this can only be set for the parent command for subcommands.
```
Comment (EmreTech) [inline nextcord/application_command.py:2749]:
```suggestion
            only usable in NSFW channels.

            .. note::

                Due to a discord limitation, this can only be set for the parent command for subcommands.

```

File: nextcord/application_command.py [hunk 42]
```diff
@@ -3043,6 +3078,13 @@ def slash_command(
         Permission(s) required to use the command. Inputting ``8`` or ``Permissions(administrator=True)`` for
         example will only allow Administrators to use the command. If set to 0, nobody will be able to use it by
         default. Server owners CAN override the permission requirements.
+    nsfw: :class:`bool`
+        If the command should be usable in NSFW channels or not. Setting to ``True`` will make the command
+        only usable in NSFW channels.
+        .. note::
+            Due to a discord limitation, this can only be set for the parent command for subcommands.
```
Comment (EmreTech) [inline nextcord/application_command.py:3085]:
```suggestion
        only usable in NSFW channels.

        .. note::

            Due to a discord limitation, this can only be set for the parent command for subcommands.

```

File: nextcord/client.py [hunk 43]
```diff
@@ -2717,6 +2731,13 @@ def slash_command(
             Permission(s) required to use the command. Inputting ``8`` or ``Permissions(administrator=True)`` for
             example will only allow Administrators to use the command. If set to 0, nobody will be able to use it by
             default. Server owners CAN override the permission requirements.
+        nsfw: :class:`bool`
+            If the command should be usable in NSFW channels or not. Setting to ``True`` will make the command
+            only usable in NSFW channels.
+            .. note::
+                Due to a discord limitation, this can only be set for the parent command for subcommands.
```
Comment (EmreTech) [inline nextcord/client.py:2738]:
```suggestion
            only usable in NSFW channels.

            .. note::

                Due to a discord limitation, this can only be set for the parent command for subcommands.

```

**Review by EmreTech — COMMENTED** [review]
Sorry I had to remake these suggestions because GitHub does not support having multiple suggestions on the same line 😅

File: nextcord/application_command.py [hunk 44]
```diff
@@ -2730,6 +2742,16 @@ def __init__(
...
             default. Server owners CAN override the permission requirements.
+        nsfw: :class:`bool`
+            If the command should be usable in NSFW channels or not. Setting to ``True`` will make the command
+            only usable in NSFW channels.
+
+            .. note::
+
+                Due to a discord limitation, this can only be set for the parent command for subcommands.
+
+
```
Comment (EmreTech) [inline nextcord/application_command.py:2753]:
```suggestion

```

File: nextcord/application_command.py [hunk 45]
```diff
@@ -3043,6 +3081,16 @@ def slash_command(
...
         default. Server owners CAN override the permission requirements.
+    nsfw: :class:`bool`
+        If the command should be usable in NSFW channels or not. Setting to ``True`` will make the command
+        only usable in NSFW channels.
+
+        .. note::
+
+            Due to a discord limitation, this can only be set for the parent command for subcommands.
+
+
```
Comment (EmreTech) [inline nextcord/application_command.py:3092]:
```suggestion

```

File: nextcord/client.py [hunk 46]
```diff
@@ -2717,6 +2731,16 @@ def slash_command(
...
             default. Server owners CAN override the permission requirements.
+        nsfw: :class:`bool`
+            If the command should be usable in NSFW channels or not. Setting to ``True`` will make the command
+            only usable in NSFW channels.
+
+            .. note::
+
+                Due to a discord limitation, this can only be set for the parent command for subcommands.
+
+
```
Comment (EmreTech) [inline nextcord/client.py:2742]:
```suggestion

```

**Review by EmreTech — COMMENTED** [review]
Apologies, I formatted incorrectly again 😅

File: nextcord/application_command.py [hunk 47]
```diff
@@ -1914,6 +1915,11 @@ def __init__(
             Permission(s) required to use the command. Inputting ``8`` or ``Permissions(administrator=True)`` for
             example will only allow Administrators to use the command. If set to 0, nobody will be able to use it by
             default. Server owners CAN override the permission requirements.
+        nsfw: :class:`bool`
```
Comment (ooliver1) [inline nextcord/application_command.py:1918]:
```suggestion
        nsfw: Optional[:class:`bool`]
```

File: nextcord/application_command.py [hunk 48]
```diff
@@ -1914,6 +1915,11 @@ def __init__(
...
+            If the command should be usable in NSFW channels or not. Setting to ``True`` will make the command
```
Comment (ooliver1) [inline nextcord/application_command.py:1919]:
```suggestion
            If the command should be solely usable in NSFW channels or not. Setting to ``True`` will make the command
```
Since this implies non age-restricted commands cant be used in restricted channels.

File: nextcord/application_command.py (hunk 10 above)
Comment (ooliver1) [reply nextcord/application_command.py:2146]: I think that at least `self.nsfw` should explicitly check `None` just to be sure.
```suggestion
        if self.nsfw is not None:
```

File: nextcord/application_command.py [hunk 49]
```diff
@@ -2730,6 +2742,15 @@ def __init__(
             Permission(s) required to use the command. Inputting ``8`` or ``Permissions(administrator=True)`` for
             example will only allow Administrators to use the command. If set to 0, nobody will be able to use it by
             default. Server owners CAN override the permission requirements.
+        nsfw: :class:`bool`
```
Comment (ooliver1) [inline nextcord/application_command.py:2745]:
```suggestion
        nsfw: Optional[:class:`bool`]
```

File: nextcord/application_command.py [hunk 50]
```diff
@@ -2730,6 +2742,15 @@ def __init__(
...
+            If the command should be usable in NSFW channels or not. Setting to ``True`` will make the command
```
Comment (ooliver1) [inline nextcord/application_command.py:2746]:
```suggestion
            If the command should be usable solely in NSFW channels or not. Setting to ``True`` will make the command
```

File: nextcord/application_command.py [hunk 51]
```diff
@@ -2889,6 +2912,11 @@ def __init__(
             Permission(s) required to use the command. Inputting ``8`` or ``Permissions(administrator=True)`` for
             example will only allow Administrators to use the command. If set to 0, nobody will be able to use it by
             default. Server owners CAN override the permission requirements.
+        nsfw: :class:`bool`
```
Comment (ooliver1) [inline nextcord/application_command.py:2915]:
```suggestion
        nsfw: Optional[:class:`bool`]
```

File: nextcord/application_command.py [hunk 52]
```diff
@@ -2889,6 +2912,11 @@ def __init__(
...
+            If the command should be usable in NSFW channels or not. Setting to ``True`` will make the command
```
Comment (ooliver1) [inline nextcord/application_command.py:2916]:
```suggestion
            If the command should be usable solely in NSFW channels or not. Setting to ``True`` will make the command
```

File: nextcord/application_command.py [hunk 53]
```diff
@@ -2965,6 +2995,11 @@ def __init__(
...
```
Comment (ooliver1) [inline nextcord/application_command.py:2999]:
```suggestion
        nsfw: Optional[:class:`bool`]
            If the command should be usable solely in NSFW channels or not. Setting to ``True`` will make the command
```

File: nextcord/client.py [hunk 54]
```diff
@@ -2611,6 +2612,11 @@ def user_command(
             Permission(s) required to use the command. Inputting ``8`` or ``Permissions(administrator=True)`` for
             example will only allow Administrators to use the command. If set to 0, nobody will be able to use it by
             default. Server owners CAN override the permission requirements.
+        nsfw: :class:`bool`
+            If the command should be usable in NSFW channels or not. Setting to ``True`` will make the command
```
Comment (ooliver1) [inline nextcord/client.py:2616]: And you get the point, I'm tired of typing that out.

**Review by ooliver1 — CHANGES_REQUESTED** [review]

**Review by EmreTech — APPROVED** [review]

**Comment by Skelmis** [comment]
Given the move to "Age Restricted" #939 would it also make sense to modify "nsfw" items in this PR to the new format

**Review by ooliver1 — COMMENTED** [review]
@alentoghostflame could you verify the defaulting to `False` instead of missing does not mess with unnecessary application command syncs?

**Review by Skelmis — APPROVED** [review]

File: nextcord/application_command.py [hunk 55]
```diff
@@ -2136,6 +2142,9 @@ def get_payload(self, guild_id: Optional[int]) -> dict:
                 #  the default (True) to ensure payload parity for comparisons.
                 ret["dm_permission"] = True
 
+        if self.nsfw:
+            ret["nsfw"] = self.nsfw
```
Comment (alentoghostflame) [inline nextcord/application_command.py:2146]: To prevent commands from needlessly resyncing, `ret["nsfw"]` must be set due to the payload Discord currently gives us.
```suggestion
        ret["nsfw"] = self.nsfw
```
If you want further details about how/why this is necessary, \@mention me in Discord.

**Review by alentoghostflame — CHANGES_REQUESTED** [review]

**Review by ooliver1 — APPROVED** [review]
